Mathlib/RingTheory/MvPowerSeries/Order.lean

```diff
@@ -0,0 +1,488 @@
+    rw [coe_toNat_eq_self]; exact ne_top_of_lt H
+  obtain ⟨d, hfd, hd⟩ := exists_coeff_ne_zero_and_weightedOrder w f this
+  rw [← hd, Nat.cast_lt] at H
+  exact hfd (h d H)
+
+/-- The order of a formal power series is at least `n` if
+the `d`th coefficient is `0` for all `d` such that `weight w d < n`.-/
+theorem le_weightedOrder {f : MvPowerSeries σ R} {n : ℕ∞}
+    (h : ∀ d : σ →₀ ℕ, ↑(weight w d) < n → coeff R d f = 0) : n ≤ f.weightedOrder w := by
+  cases n
+  · rw [top_le_iff, weightedOrder_eq_top_iff]
+    ext d; exact h d (coe_lt_top _)
+  · apply nat_le_weightedOrder;
+    simpa only [ENat.some_eq_coe, Nat.cast_lt] using h
+
+/-- The order of a formal power series is exactly `n` if and only if some coefficient of weight `n`
+is nonzero, and the `d`th coefficient is `0` for all `d` such that `weight w d < n`.-/
+theorem weightedOrder_eq_nat {f : MvPowerSeries σ R} {n : ℕ} :
+    f.weightedOrder w = n ↔
+      (∃ d, coeff R d f ≠ 0 ∧ weight w d = n) ∧ ∀ d, weight w d < n → coeff R d f = 0 := by
+  constructor
+  · intro h
+    obtain ⟨d, hd⟩ := f.exists_coeff_ne_zero_and_weightedOrder w (by simp only [h, toNat_coe])
+    exact ⟨⟨d, by simpa [h, Nat.cast_inj, ne_eq] using hd⟩,
+      fun e he ↦ f.coeff_eq_zero_of_lt_weightedOrder w (by simp only [h, Nat.cast_lt, he])⟩
+  · rintro ⟨⟨d, hd', hd⟩, h⟩
+    exact le_antisymm (hd.symm ▸ f.weightedOrder_le w hd') (nat_le_weightedOrder w h)
+
+/-- The weighted_order of the monomial `a*X^d` is infinite if `a = 0` and `weight w d` otherwise.-/
+theorem weightedOrder_monomial (d : σ →₀ ℕ) (a : R) [Decidable (a = 0)] :
+    weightedOrder w (monomial R d a) = if a = 0 then (⊤ : ℕ∞) else weight w d := by
+  classical
+  split_ifs with h
+  · rw [h, weightedOrder_eq_top_iff, LinearMap.map_zero]
+  · rw [weightedOrder_eq_nat]
+    constructor
+    · use d
+      simp only [coeff_monomial_same, ne_eq, h, not_false_eq_true, and_self]
+    · intro b hb
+      rw [coeff_monomial, if_neg]
+      intro h
+      simp only [h, lt_self_iff_false] at hb
+
+/-- The order of the monomial `a*X^n` is `n` if `a ≠ 0`.-/
+theorem weightedOrder_monomial_of_ne_zero (d : σ →₀ ℕ) (a : R) (h : a ≠ 0) :
+    weightedOrder w (monomial R d a) = weight w d := by
+  classical
+  rw [weightedOrder_monomial, if_neg h]
+
+
+/-- The order of the sum of two formal power series is at least the minimum of their orders.-/
+theorem min_weightedOrder_le_add (f g : MvPowerSeries σ R) :
+    min (f.weightedOrder w) (g.weightedOrder w) ≤ (f + g).weightedOrder w := by
+  apply le_weightedOrder w
+  simp (config := { contextual := true }) only
+    [coeff_eq_zero_of_lt_weightedOrder w, lt_min_iff, map_add, add_zero,
+      eq_self_iff_true, imp_true_iff]
+
+private theorem weightedOrder_add_of_weightedOrder_lt.aux {f g : MvPowerSeries σ R}
+    (H : f.weightedOrder w < g.weightedOrder w) :
+    (f + g).weightedOrder w = f.weightedOrder w := by
+  obtain ⟨n, hn⟩ := ne_top_iff_exists.mp (ne_top_of_lt H)
+  erw [← hn, weightedOrder_eq_nat]
```

Comment:
Why do you need `erw`?

Reply:
because `rw` doesn't work, even if I provide all arguments I can think of.

Reply:
On the other hand, I noticed that `ne_top_iff_exists` furnishes something in `WithTop Nat`, while `weightedOrder_eq_nat` expected something in `Nat\infty`, so specifying the type of `hn`on line 250 was enough to have `rw` work.